Fragment banners

tbielawa opened the pull request:
I think wrapping fragment banners around the included pieces in a generated library file would help improve readability. Included is a commit which updates the `generate.py` file, as well as an example of what happens after you run it

@ashcrow  @kwoodson @sdodson @abutcher @mtnbikenc I would love to hear your thoughts on this. I'm not married to the idea, I'm just pitchin' it because I like it. Happy to cancel the PR if nobody sees much value in this :+1:

tbielawa commented:
If we get consensus on this I'd backout the commit that shows the diff and let us work that out in other pull requests before we merge this one.

ashcrow commented:
Will review shortly.

kwoodson commented:
@tbielawa,

If this makes the code more readable from a reviewer and from external sources then I'm +1.

ashcrow commented on roles/lib_openshift/src/generate.py:
```diff
@@ -27,19 +27,47 @@ def parse_args():
+    }
+    annotation = side_msg[side] % fragment_path
+
+    banner = """
```
nit: A smaller, but highly visible banner is another option calling pit generated code blocks withoit as much real estate. Something like: ```# - - -  XXX Begin include fragment - - -```.

tbielawa replied:
ACK

ashcrow commented on roles/lib_openshift/src/generate.py:
```diff
@@ -27,19 +27,47 @@ def parse_args():
+######################################################################
+""".format(annotation)
+
+    if ("generated" not in fragment_path) and ("license" not in fragment_path):
```
Code documentation why these are skipped please.

ashcrow replied:
@tbielawa explained this was to avoid wrapping generated code and messing with shebangs.

ashcrow commented on roles/lib_utils/src/generate.py:
```diff
@@ -27,19 +27,47 @@ def parse_args():
     return parser.parse_args()
 
 
+def fragment_banner(fragment_path, side, data):
```
Could this and the above library be combined or is it required to have two copies due to ansible restrictions?

tbielawa replied:
I'm not sure why it's duplicated in both roles. I could probably even be, dare I say it, symlinked.

ashcrow replied:
This is still split but it can be kept and updated later as tech debt.

ashcrow commented:
aos-ci-test

ashcrow commented:
@tbielawa note a rebase is needed.

tbielawa commented on roles/lib_openshift/src/generate.py:
```diff
@@ -27,19 +27,47 @@ def parse_args():
+def fragment_banner(fragment_path, side, data):
+    """Generate a banner to wrap around file fragments
+
+* :param fragment_path: A path to a module fragment
```
Moving to all-in-one format, `:param <TYPE> <NAME>:`

tbielawa commented on roles/lib_openshift/src/generate.py:
```diff
@@ -27,19 +27,47 @@ def parse_args():
+* :param data: A StringIO object to write the banner to
+"""
+    side_msg = {
+        "header": "Begin included fragment: %s",
```
ACK.

tbielawa commented:
Rebased onto latest master.

Addressed code style and documentation stuff from @ashcrow 

Slimmed the banner down to a single line

```
# -*- -*- -*- Begin included fragment: class/yedit.py -*- -*- -*-
...
...
# -*- -*- -*- End included fragment: class/yedit.py -*- -*- -*-
```

tbielawa commented:
aos-ci-test

ashcrow commented on roles/lib_openshift/src/generate.py:
```diff
@@ -27,19 +27,53 @@ def parse_args():
+    annotation = side_msg[side].format(fragment_path)
+
+    banner = """
+# -*- -*- -*- {} -*- -*- -*-
```
:+1:

ashcrow commented on roles/lib_openshift/src/generate.py:
```diff
@@ -27,19 +27,53 @@ def parse_args():
+# -*- -*- -*- {} -*- -*- -*-
+""".format(annotation)
+
+    # Why skip?
```
:+1:

ashcrow commented on roles/lib_openshift/src/generate.py:
```diff
@@ -27,19 +27,53 @@ def parse_args():
-    '''generate the source code for the ansible modules'''
+    '''generate the source code for the ansible modules
+
+* :param parts: An array of paths to module fragments
```
The ``*`` is still unneeded:

```python
"""
...
    :param parts: An array of paths to module fragments.
"""
```

but it's not worth blocking on :smile:.

tbielawa commented:
@ashcrow updated the `:param:` that had the extra star. Please to review and approve :D

tbielawa commented:
aos-ci-test

tbielawa commented:
Whoops. Looks like some lint got in there.

Also looks like I can't pass the tests unless the artifacts generated from the fragments match the expected output.

tbielawa commented:
Rebased, fixed lint, generated artifacts from fragments. blablabla.

tbielawa commented:
aos-ci-test

tbielawa commented:
What in the heck. It seems like the `mccabe` python dep (*McCabe checker, plugin for flake8*) is failing.

```
$ /home/tbielawa/rhat/os/openshift-ansible/utils/.tox/py27-flake8/bin/python setup.py flake8
Traceback (most recent call last):
  File "setup.py", line 77, in <module>
    'oo-install=ooinstall.cli_installer:cli',
  File "/usr/lib64/python2.7/distutils/core.py", line 137, in setup
    ok = dist.parse_command_line()
  File "/home/tbielawa/rhat/os/openshift-ansible/utils/.tox/py27-flake8/lib/python2.7/site-packages/setuptools/dist.py", line 358, in parse_command_line
    result = _Distribution.parse_command_line(self)
  File "/usr/lib64/python2.7/distutils/dist.py", line 467, in parse_command_line
    args = self._parse_command_opts(parser, args)
  File "/home/tbielawa/rhat/os/openshift-ansible/utils/.tox/py27-flake8/lib/python2.7/site-packages/setuptools/dist.py", line 669, in _parse_command_opts
    nargs = _Distribution._parse_command_opts(self, parser, args)
  File "/usr/lib64/python2.7/distutils/dist.py", line 523, in _parse_command_opts
    cmd_class = self.get_command_class(command)
  File "/home/tbielawa/rhat/os/openshift-ansible/utils/.tox/py27-flake8/lib/python2.7/site-packages/setuptools/dist.py", line 489, in get_command_class
    ep.require(installer=self.fetch_build_egg)
  File "/home/tbielawa/rhat/os/openshift-ansible/utils/.tox/py27-flake8/lib/python2.7/site-packages/pkg_resources/__init__.py", line 2306, in require
    items = working_set.resolve(reqs, env, installer)
  File "/home/tbielawa/rhat/os/openshift-ansible/utils/.tox/py27-flake8/lib/python2.7/site-packages/pkg_resources/__init__.py", line 853, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
pkg_resources.VersionConflict: (mccabe 0.6.0 (/home/tbielawa/rhat/os/openshift-ansible/utils/.tox/py27-flake8/lib/python2.7/site-packages), Requirement.parse('mccabe<0.6.0,>=0.5.0'))
```

Check it out on PyPi: https://pypi.python.org/pypi/mccabe/ **Version 0.6.0** was released today, 2017-01-23

```
0.6.0 - 2017-01-23
Add support for Python 3.6
Fix handling for missing statement types
```

Back to our code, I look inside of `oo-installenv/lib/python2.7/site-packages/flake8-3.2.1.dist-info/METADATA` and find this:

```
Requires-Dist: mccabe >= 0.5.0, < 0.6.0
```

So it seems that `flake8` is at the root of the error. The deps listed there are prohibiting installation of the latest mccabe release. What's still available in pip?

```
$ pip install mccabe==99999
Collecting mccabe==99999
  Could not find a version that satisfies the requirement mccabe==99999 (from versions: 0.2, 0.2.1, 0.3, 0.3.1, 0.4.0, 0.5.0, 0.5.1, 0.5.2, 0.5.3, 0.6.0)
No matching distribution found for mccabe==99999
```

**HARD CODING** the `mccabe` version to the latest in the 0.5.x series allows the tests to complete

```diff
diff --git a/utils/test-requirements.txt b/utils/test-requirements.txt
index f6a7bde..427b0e3 100644
--- a/utils/test-requirements.txt
+++ b/utils/test-requirements.txt
@@ -13,3 +13,4 @@ pyOpenSSL
 yamllint
 tox
 detox
+mccabe==0.5.3
```

```
<utils/> $ make ci
. oo-installenv/bin/activate && detox
py27-flake8 create: /home/tbielawa/rhat/os/openshift-ansible/utils/.tox/py27-flake8
py27-unit create: /home/tbielawa/rhat/os/openshift-ansible/utils/.tox/py27-unit
...
py35-flake8 runtests: PYTHONHASHSEED='273850839'
py35-flake8 runtests: commands[0] | python setup.py flake8
__________ summary __________
  py27-flake8: commands succeeded
  py27-unit: commands succeeded
  py27-pylint: commands succeeded
  py35-flake8: commands succeeded
  py35-unit: commands succeeded
  py35-pylint: commands succeeded
  congratulations :) 

##################################################################################
VIEW CODE COVERAGE REPORT WITH 'xdg-open cover/index.html' or run 'make viewcover'
To clean your test environment run 'make clean'
Other targets you may run with 'make': 'ci-pylint', 'ci-tox', 'ci-unittests', 'ci-flake8'
```

tbielawa commented:
Pinned the mccabe dep to 0.5.3 to work around the issue described in the previous comment.

tbielawa commented:
aos-ci-test

ashcrow commented:
@tbielawa [fixes are upstream](https://gitlab.com/pycqa/flake8/blob/master/setup.py#L22) and should be part of the next release.

ashcrow commented:
Upstream issue opened [here](https://gitlab.com/pycqa/flake8/issues/301).

tbielawa commented:
aos-ci-test

tbielawa commented:
Rebased on top of latest master which includes the maccabe dependency work-around.

tbielawa commented:
ugh. force pushed with a commit msg update to try and get CI to play along.

tbielawa commented:
aos-ci-test
